### PR TITLE
FIX: Smart list jankiness in Chrome and fix for Firefox

### PR DESCRIPTION
**Pull request by martin-brennan**
Last week I disabled smart lists in Firefox in 2ab4913d1343b1a9b8fd131b142f89d88480fbc9.

This week the same issue presented itself in Chrome. Turns out,
the list modification was still not firing at the right time
in the event chain. I investigated and it looks as though
`beforeinput` is a better fit, since:

> This allows web apps to override text edit behavior before the browser
modifies the DOM tree, and provides more control over input events to
improve performance.

c.f. https://developer.mozilla.org/en-US/docs/Web/API/Element/beforeinput_event
and https://webkit.org/blog/7358/enhanced-editing-with-input-events/
and https://www.w3.org/TR/uievents/#events-keyboard-event-order

The order of keyboard events is `keydown` -> `beforeinput` -> `input` -> `keyup`

I changed to detect the event type of `insertLineBreak` which is
not always consistently true in `input` events. If it's true when
`beforeinput` is fired then we go ahead with the smart list when
`input` fires.

### CODE REVIEW

**Comment by martin-brennan** [comment]
Hmm will just have to fix the tests here, they are using `keydown`...hopefully I can simulate `beforeinput` somehow :sweat_smile:

File: app/assets/javascripts/discourse/tests/integration/components/d-editor-test.js [hunk 1]
```diff
@@ -989,16 +976,33 @@ third line`
...
+  function triggerEnter(textarea) {
+    let event = new Event("beforeinput");
+    event.inputType = "insertLineBreak";
+    textarea.dispatchEvent(event);
+
+    event = new Event("input");
+    event.inputType = "insertText";
+    event.data = "\n";
+    textarea.dispatchEvent(event);
+  }
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/tests/integration/components/d-editor-test.js:994]: Should this use: https://github.com/emberjs/ember-test-helpers/blob/master/API.md#triggerevent

So we can do async/await on triggerEnter

**Review by jjaffeux — APPROVED** [review]

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 2]
```diff
@@ -351,6 +363,26 @@ export default Component.extend(TextareaTextManipulation, {
...
+  @bind
+  onBeforeInputSmartList(event) {
+    // This inputType is much more consistently fired in `beforeinput`
+    // rather than `input`.
+    const isNewLine = event.inputType === "insertLineBreak";
+    if (isNewLine) {
+      this.handleSmartListAutocomplete = true;
+    } else {
+      this.handleSmartListAutocomplete = false;
+    }
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/d-editor.js:375]:
```suggestion
    this.handleSmartListAutocomplete = isNewLine;
```

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/d-editor.js:375]: Don't come in here with your fancy last minute refactors CVX :P

**Pull request merged**